src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -456,6 +855,9 @@ public int OnItemChosen(string pszTitle, string pszPath)
 
                 var expansion = EditorAdaptersFactoryService.GetBufferAdapter(textViewModel.DataBuffer) as IVsExpansion;
                 earlyEndExpansionHappened = false;
+                _symbols = default;
+                _symbol = null;
+                _arguments = _arguments.Clear();
                 hr = expansion.InsertNamedExpansion(pszTitle, pszPath, textSpan, this, LanguageServiceGuid, fShowDisambiguationUI: 0, pSession: out ExpansionSession);
 
                 if (earlyEndExpansionHappened)
```

Comment:
so, it's late over here and i skipped to this point.  I'll go over tihs file again tomorrow.  But i def believe this needs a little TLC overall.  I wholeheartedly approve of the feature, and i think it's fine for there to be a connection between sighelp and snippets.  However, i think we need a stronger way to do this as the current approach seems a little too hacky to me.

Also, having like 10 cases of "clear out this state" is a pretty bad smell :)   i personally think we should conside trying ot extract out hte mutable 'current state' aspect of this type into it's own container that can be accessed/cleared independently.  However, i don't think that is a requirement.  At the very least though cleanup would be nice here.

--

Finally, while possibly a pain, i think we should discuss if the 'async mvc' pattern would be appropriate here.  What i mean by that is to follow a pattern more closely like what we did with completion/sighelp.  Namely, we have a UI affinitized component that is listening for user interactions, and this chains work to a BG task queue that actually then goes and processes the user interactions, doing async work as necessary, and updating hte UI when the computations it does to its model complete.  This would likely help address a lot of gnarly UI-thread+async-work goop that i see all around.

This would be a substantial amount of effort to do.  but i think overall it would make things much clearer and would both delineate responsibility, as well as make it clear where the UI portions of this code should be, vs the potentially-expensive async work.  

Tagging @jasonmalinowski for his thoughts here as well.  IMO it's been a successful pattern for both completion and sighelp.  And given that tihs is effectively merging snippets+sig-help i think it would make a lot of sense to be used here.

Reply:
The snippets platform/API is being completely rewritten. Aside from separation of `ArgumentProvider` into its own type (already done), I'm not planning to try and hide the limitations of the current snippets API in the command handler or client. Doing so will only make it harder to adopt the new asynchronous platform API when it becomes available. The goals of the current approach are:

1. Contain the Insert Full Method Call paths so it's clear which ugly code handles normal snippets and which ugly code handles the new feature
2. Avoid having Insert Full Method Call influence runtime behavior in cases where it is not being used (and especially when it is not available)
3. Make sure partners that integrate with this feature do so exclusively through `ArgumentProvider` so we are not blocked from rewriting the snippets integration when a new API becomes avaliable.

Reply:
i'm going to resolve this as the changes have been substantial and have substanively addressed the meat of my concerns.